File: 105. Construct Binary Tree from Preorder and Inorder Traversal.cpp
class Solution {
public:
    TreeNode* build(queue<int>& preOrder, vector<int>& inorder, unordered_map<int, int>& hm, int lastInd) {
        if (preOrder.empty()) return nullptr;

        int val = preOrder.front();
        int index = hm[val];

        if (index > lastInd) return nullptr;

        preOrder.pop();
        TreeNode* curr = new TreeNode(val);
        curr->left = build(preOrder, inorder, hm, index);
        curr->right = build(preOrder, inorder, hm, lastInd);

        return curr;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        unordered_map<int, int> hm;
        for (int i = 0; i < (int)inorder.size(); i++) {
            hm[inorder[i]] = i;
        }

        queue<int> q;
        for (int node : preorder) q.push(node);

        return build(q, inorder, hm, inorder.size() - 1);
    }
};

File: 105. Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int index = 0;
    public TreeNode build(int[] preorder, HashMap<Integer, Integer> hm, int left, int right){
        if(left > right) return null;
        TreeNode node = new TreeNode(preorder[index]);
        index++;

        int rootIndex = hm.get(node.val);
        node.left = build(preorder, hm, left, rootIndex - 1);
        node.right = build(preorder, hm, rootIndex + 1, right);

        return node;
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        HashMap<Integer, Integer> hm = new HashMap<>();
        for(int i = 0; i < inorder.length; i++){
            hm.put(inorder[i], i);
        }
        TreeNode root = build(preorder, hm, 0, inorder.length - 1);
        return root;
    }
}